File: Common/Actor/src/common/actor/ControlableActor.java
package common.actor;

import javafx.scene.canvas.Canvas;
import javafx.scene.input.KeyEvent;
import java.util.HashMap;

abstract public class ControlableActor extends Actor {

    protected ControlableActor( final String pictureFileName,
                                final double x,
                                final double y,
                                final HashMap<String, Direction> keymap ) {
        super( pictureFileName, x, y );
        setKeyMap(keymap);
    }

    protected ControlableActor( final String[] pictureFileName,
                                final double x,
                                final double y,
                                final HashMap<String, Direction> keymap,
                                final int delay ) {
        super( x, y, delay, pictureFileName );
        setKeyMap(keymap);
    }

    /**
    * Checks Key Released and Pressed Events.
    * */
    public void move( final KeyEvent keyEvent ) {

        final String keyName   = keyEvent.getCode().getName();
        final String eventName = keyEvent.getEventType().getName();

        if ( eventName.equals( "KEY_PRESSED" ) ) {
            this.movement.onKeyPressed( keyName );
        } else if ( eventName.equals( "KEY_RELEASED" ) ) {
            this.movement.onKeyReleased( keyName );
        }
    }

    @Override
    public void draw( Canvas canvas ) {
        this.draw( canvas, canvas.getWidth(), canvas.getHeight() );
    }

    @Override
    public void draw( Canvas gc, final double canvas_width, final double canvas_height ) {
        final double[] temp = this.calculateNewPosFromInput();
        super.draw( gc, canvas_width, canvas_height, temp[ 0 ], temp[ 1 ] );
    }

    private double[] calculateNewPosFromInput() {
        final double[] xyTuple  = new double[ 2 ];
        final double   velocity = movement.velocity;
        movement.getDirections().stream()
            .filter(movement::isHoldDown)
            .forEach(direction -> {
                final double[] temp = calculateDirectedSpeed( direction, velocity );
                xyTuple[ 0 ] = temp[ 0 ];
                xyTuple[ 1 ] = temp[ 1 ];
            });
        return xyTuple;
    }

    protected double[] calculateDirectedSpeed( final Direction direction, final double movement_speed ) {
        final double[] xyTuple = new double[ 2 ];
        if ( direction == Direction.Down ) {
            xyTuple[ 1 ] += movement_speed;
        }
        if ( direction == Direction.Up ) {
            xyTuple[ 1 ] += -movement_speed;
        }
        if ( direction == Direction.Left ) {
            xyTuple[ 0 ] += -movement_speed;
        }
        if ( direction == Direction.Right ) {
            xyTuple[ 0 ] += movement_speed;
        }
        return xyTuple;
    }
}

File: Common/Actor/src/common/actor/ControlableActor.kt
package common.actor

import javafx.scene.canvas.Canvas
import javafx.scene.input.KeyEvent

abstract class ControlableActor : Actor {
    protected constructor(pictureFileName: String,
                          x: Double,
                          y: Double,
                          keymap: Map<String, Direction>) : super(pictureFileName, x, y) {
        this.movement.keymap = (keymap)
    }

    protected constructor(pictureFileNames: Array<String>,
                          x: Double,
                          y: Double,
                          keymap: Map<String, Direction>,
                          delay: Int) : super(pictureFileNames.toList(), x = x, y = y, delay = delay) {
        this.movement.keymap = (keymap)
    }

    private var facingDirection: Direction = Direction.Left

    /**
     * Checks Key Released and Pressed Events.
     */
    fun move(keyEvent: KeyEvent) {
        val keyName = keyEvent.code.getName()
        when (keyEvent.eventType.name) {
            "KEY_PRESSED" -> { movement.onKeyPressed(keyName) }
            "KEY_RELEASED" -> { movement.onKeyReleased(keyName) }
            else -> {}
        }
    }

    override fun draw(canvas: Canvas) {
        val directedSpeed = movement.holdDown.keys
                .filter { direction: Direction -> movement.holdDown[direction]!! }
                .map { direction: Direction -> calculateDirectedSpeed(direction, movement.velocity) }
                .firstOrNull()
        if (directedSpeed != null) {
            super.draw(canvas, directedSpeed[0], directedSpeed[1])
        } else {
            super.draw(canvas)
        }
    }

    protected open fun calculateDirectedSpeed(direction: Direction, movementSpeed: Double): DoubleArray {
        val xyTuple = DoubleArray(2)
        when (direction) {
            Direction.Right -> xyTuple[0] = movementSpeed
            Direction.Left -> xyTuple[0] = -movementSpeed
            Direction.Down -> xyTuple[1] = movementSpeed
            Direction.Up -> xyTuple[1] = -movementSpeed
            Direction.Non -> {}
        }
        return xyTuple
    }

    protected fun changeFacingDirection(directionChange: DoubleArray) {
        mirrorImageOnVerticalAxis(horizontalDirectionChange = directionChange[0])
        mirrorImageOnHorizontalAxis(verticalDirectionChange = directionChange[1])
    }

    private fun mirrorImageOnVerticalAxis(horizontalDirectionChange: Double, invertImage: Boolean = true) {
        if (horizontalDirectionChange > 0 && this.facingDirection != Direction.Right) {
            this.facingDirection = Direction.Right
            if (invertImage && this.scaleWidth > 0) {
                this.scaleImageWidth(-1.0)
            }
        } else if (horizontalDirectionChange < 0 && this.facingDirection != Direction.Left) {
            this.facingDirection = Direction.Left
            if (invertImage && this.scaleWidth < 0) {
                this.scaleImageWidth(-1.0)
            }
        }
    }

    private fun mirrorImageOnHorizontalAxis(verticalDirectionChange: Double, invertImage: Boolean = false) {
        if (verticalDirectionChange > 0 && this.facingDirection != Direction.Down) {
            this.facingDirection = Direction.Down
            if (invertImage && this.scaleHeight < 0) {
                this.scaleImageHeight(-1.0)
            }
        } else if (verticalDirectionChange < 0 && this.facingDirection != Direction.Up) {
            this.facingDirection = Direction.Up
            if (invertImage && this.scaleHeight < 0) {
                this.scaleImageHeight(-1.0)
            }
        }
    }
}
